Stop serializing CGFloat

charliewolfe opened the pull request:
#### 697bceba2d974bed1415abe4ee1efbfc4236c870
<pre>
Stop serializing CGFloat
<a href="https://bugs.webkit.org/show_bug.cgi?id=295571">https://bugs.webkit.org/show_bug.cgi?id=295571</a>
<a href="https://rdar.apple.com/155295427">rdar://155295427</a>

Reviewed by Alex Christensen.

In the watchOS 26 beta, legacy apps have a UI process that runs as arm64_32 but has arm64 child
processes. CGFloat serialization is not symmetric between these architectures, so we need to remove it.

Add DoubleGeometry to replace CoreGraphics types using CGFloat without losing precision.
GlyphBufferAdvance is already constructed using floats, so no precision will be lost when switching to
FloatSize.

* LayoutTests/ipc/serialized-type-info.html:
* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::setLineDash):
(WebKit::RemoteDisplayListRecorder::drawGlyphs):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheDecomposedGlyphs):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(class_template_headers):
* Source/WebKit/Shared/Cocoa/CoreIPCNSShadow.h:
(WebKit::CoreIPCNSShadow::CoreIPCNSShadow):
* Source/WebKit/Shared/Cocoa/CoreIPCNSShadow.mm:
(WebKit::CoreIPCNSShadow::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSShadow.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm:
(WebKit::CoreIPCNSValue::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.serialization.in:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionTab.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionTabParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionWindow.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionWindowParameters.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/graphics/DoubleGeometry.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSShadow.h.
(WebKit::DoublePoint::DoublePoint):
(WebKit::DoublePoint::toCG const):
(WebKit::DoubleSize::DoubleSize):
(WebKit::DoubleSize::toCG const):
(WebKit::DoubleRect::DoubleRect):
(WebKit::DoubleRect::toCG const):
* Source/WebKit/Shared/graphics/DoubleGeometry.serialization.in: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSValue.serialization.in.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWindowsCocoa.mm:
(WebKit::WebExtensionContext::windowsCreate):
(WebKit::WebExtensionContext::windowsUpdate):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::toWebAPI):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::setLineDash):
(WebKit::RemoteDisplayListRecorderProxy::drawGlyphsImmediate):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheDecomposedGlyphs):

Canonical link: <a href="https://commits.webkit.org/297177@main">https://commits.webkit.org/297177@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/911567bf2e683c4b999338adf78da9e2fc6d76dc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110777 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30440 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20872 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116806 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112740 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31120 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39026 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/84231 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113725 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/99759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64672 "Found 1 new API test failure: /WPE/TestCookieManager:/webkit/WebKitCookieManager/replace-get-all-cookies (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/110209 "Passed tests") | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/17895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/60599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/17955 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119596 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37822 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/28110 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38198 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96026 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/93016 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/38053 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/15807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17871 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37715 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43185 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/37376 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/40714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/39083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
```diff
@@ -289,9 +289,11 @@ void RemoteDisplayListRecorder::setLineCap(LineCap lineCap)
+void RemoteDisplayListRecorder::setLineDash(FixedVector<double>&& dashArray, float dashOffset)
 {
-    context().setLineDash(dashArray, dashOffset);
+    context().setLineDash(DashArray::createWithSizeFromGenerator(dashArray.size(), [&](size_t i) {
```
I think using the span constructor would be more elegant.

achristensen07 commented on Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
```diff
@@ -404,15 +406,18 @@ void RemoteDisplayListRecorder::drawFilteredImageBuffer(std::optional<RenderingR
     }
 
-    context().drawGlyphs(*font, Vector(glyphsAdvances.span<0>()), Vector(glyphsAdvances.span<(1)>()), localAnchor, fontSmoothingMode);
+    auto advances = Vector<GlyphBufferAdvance>(glyphsAdvances.span<1>().size(), [&](size_t i) {
```
This code has become less clear.  Same in RemoteRenderingBackend::cacheDecomposedGlyphs.  Can we construct the Vector without a lambda accessor?

achristensen07 commented on Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
```diff
@@ -168,7 +168,9 @@ void RemoteDisplayListRecorderProxy::setLineCap(LineCap lineCap)
 void RemoteDisplayListRecorderProxy::setLineDash(const DashArray& array, float dashOffset)
 {
-    send(Messages::RemoteDisplayListRecorder::SetLineDash(array, dashOffset));
+    send(Messages::RemoteDisplayListRecorder::SetLineDash(FixedVector<double>::createWithSizeFromGenerator(array.size(), [&](size_t i) {
```
span constructor?

achristensen07 commented on Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
```diff
@@ -271,7 +273,10 @@ void RemoteDisplayListRecorderProxy::drawGlyphsImmediate(const Font& font, std::
     appendStateChangeItemIfNecessary();
     recordResourceUse(const_cast<Font&>(font));
-    send(Messages::RemoteDisplayListRecorder::DrawGlyphs(font.renderingResourceIdentifier(), { glyphs.data(), advances.data(), glyphs.size() }, localAnchor, smoothingMode));
+    auto advancesVector = Vector<FloatSize>(advances.size(), [&](size_t i) {
```
span constructor?

achristensen07 commented on Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
```diff
@@ -436,7 +436,10 @@ void RemoteRenderingBackendProxy::releaseFontCustomPlatformData(RenderingResourc
 void RemoteRenderingBackendProxy::cacheDecomposedGlyphs(const DecomposedGlyphs& glyphs)
 {
-    send(Messages::RemoteRenderingBackend::CacheDecomposedGlyphs({ glyphs.glyphs().data(), glyphs.advances().data(), glyphs.glyphs().size() }, glyphs.localAnchor(), glyphs.fontSmoothingMode(), glyphs.renderingResourceIdentifier()));
+    auto advancesVector = Vector<FloatSize>(glyphs.advances().size(), [&](size_t i) {
```
span constructor?

charliewolfe committed:
Stop serializing CGFloat
https://bugs.webkit.org/show_bug.cgi?id=295571
rdar://155295427

Reviewed by Alex Christensen.

In the watchOS 26 beta, legacy apps have a UI process that runs as arm64_32 but has arm64 child
processes. CGFloat serialization is not symmetric between these architectures, so we need to remove it.

Add DoubleGeometry to replace CoreGraphics types using CGFloat without losing precision.
GlyphBufferAdvance is already constructed using floats, so no precision will be lost when switching to
FloatSize.

* LayoutTests/ipc/serialized-type-info.html:
* Source/WebCore/editing/cocoa/AttributedString.h:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.cpp:
(WebKit::RemoteDisplayListRecorder::setLineDash):
(WebKit::RemoteDisplayListRecorder::drawGlyphs):
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.h:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::cacheDecomposedGlyphs):
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(class_template_headers):
* Source/WebKit/Shared/Cocoa/CoreIPCNSShadow.h:
(WebKit::CoreIPCNSShadow::CoreIPCNSShadow):
* Source/WebKit/Shared/Cocoa/CoreIPCNSShadow.mm:
(WebKit::CoreIPCNSShadow::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSShadow.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.h:
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.mm:
(WebKit::CoreIPCNSValue::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCNSValue.serialization.in:
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionTab.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionTabParameters.h:
* Source/WebKit/Shared/Extensions/WebExtensionWindow.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionWindowParameters.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/graphics/DoubleGeometry.h: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSShadow.h.
(WebKit::DoublePoint::DoublePoint):
(WebKit::DoublePoint::toCG const):
(WebKit::DoubleSize::DoubleSize):
(WebKit::DoubleSize::toCG const):
(WebKit::DoubleRect::DoubleRect):
(WebKit::DoubleRect::toCG const):
* Source/WebKit/Shared/graphics/DoubleGeometry.serialization.in: Copied from Source/WebKit/Shared/Cocoa/CoreIPCNSValue.serialization.in.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIWindowsCocoa.mm:
(WebKit::WebExtensionContext::windowsCreate):
(WebKit::WebExtensionContext::windowsUpdate):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::toWebAPI):
* Source/WebKit/WebProcess/GPU/graphics/RemoteDisplayListRecorderProxy.cpp:
(WebKit::RemoteDisplayListRecorderProxy::setLineDash):
(WebKit::RemoteDisplayListRecorderProxy::drawGlyphsImmediate):
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::cacheDecomposedGlyphs):

Canonical link: https://commits.webkit.org/297177@main
